internal/controller/drpolicy_peerclass.go

```diff
@@ -0,0 +1,590 @@
+// SPDX-FileCopyrightText: The RamenDR authors
+// SPDX-License-Identifier: Apache-2.0
+
+package controllers
+
+import (
+	"fmt"
+	"slices"
+
+	volrep "github.com/csi-addons/kubernetes-csi-addons/api/replication.storage/v1alpha1"
+	"github.com/go-logr/logr"
+	snapv1 "github.com/kubernetes-csi/external-snapshotter/client/v8/apis/volumesnapshot/v1"
+	"github.com/stolostron/multicloud-operators-foundation/pkg/apis/view/v1beta1"
+	storagev1 "k8s.io/api/storage/v1"
+
+	ramen "github.com/ramendr/ramen/api/v1alpha1"
+	"github.com/ramendr/ramen/internal/controller/util"
+)
+
+// classLists contains [storage|snapshot|replication]classes from ManagedClusters with the required ramen storageID or,
+// replicationID labels
+type classLists struct {
+	clusterID string
+	sClasses  []*storagev1.StorageClass
+	vsClasses []*snapv1.VolumeSnapshotClass
+	vrClasses []*volrep.VolumeReplicationClass
+}
+
+// peerInfo contains a single peer relationship between a PAIR of clusters for a common storageClassName across
+// these peers. This should directly translate to DRPolicy.Status.[Async|Sync] updates.
+// NOTE: storageID discussed in comments relates to the value of the label "ramendr.openshift.io/storageid" for the
+// respective class. replicationID in comments relates to the value of the label "ramendr.openshift.io/replicationid"
+// for the respective class
+type peerInfo struct {
+	// replicationID is an empty string (indicating no common VolumeReplicationClass) or the common replicationID value
+	// for the corresponding VRClass on each peer
+	replicationID string
+
+	// storageIDs is a list containing,
+	// - A single storageID if the storageClassName across the peers have the same storageID, denoting a synchronous
+	// pairing for the storageClassName
+	// -  It is a pair of storageIDs, if there exists an asynchronous pairing between the clusters, either due to
+	// a common replicationID or due to required VolumeSnapshotClasses on each cluster
+	storageIDs []string
```

Comment:
This is a bit confusing to me. Under what conditions would there be more than one `storageID` for the same `storageClassName`? I’m trying to understand: in the case of single`storageID`, then `c1` and `c2` sharing the same StorageClass makes sense. But in what case would there be more than one `storageIDs`? Is it because the StorageClass has the same name in both clusters but refers to two different storage systems?

Reply:
If SC has the same name (which is required to move PVCs across clusters without changing the PVC SC field), but is pointing to different storage systems (Async replication), they will have different storageIDs.

So if a peer has a single storageID then it is a sync peer, if it has 2 storageIDs then it is an Async peer.